### PR TITLE
[Parse] [Sema] Update confusables diagnostic to mention the character names as well

### PR DESCRIPTION
**Pull request by theblixguy**
This is a follow up to #33047.

When we encounter confusable characters, we emit a diagnostic note to let the user know about it. This PR improves the diagnostic to include the names of the confusable and expected characters. This is quite helpful in situations where both the confusable and expected characters look identical to the user and the character names help provide more clarity and information.

For example, previously we would say:

```swift
print("A"); print("B") // identifier ';' contains possibly confused characters; did you mean to use ';'?
```
(and notice that both characters _look identical_)

and now we say:

```swift
print("A"); print("B") // identifier ';' (Greek Question Mark) looks similar to ';' (Semicolon); did you mean ';' (Semicolon)?
```

The character names are only present when there is a single code point (like above, which is a tailored diagnostic that I have added). Otherwise, they are not shown as there could be various code points and it could make the diagnostic a lot longer.

The Rust language also has a similar [diagnostic](https://github.com/rust-lang/rust/blob/master/src/test/ui/parser/unicode-chars.stderr) which mentions the character names, from which I have taken inspiration.

### CODE REVIEW

**Comment by rintaro** [comment]
Whoa, I just noticed that we have a generator of `include/swift/Parse/Confusables.def`.

https://github.com/apple/swift/blob/master/utils/generate_confusables.py
https://github.com/apple/swift/blob/master/utils/UnicodeData/confusables.txt

Could you update them and generate the file from it?

**Comment by theblixguy** [comment]
@rintaro Yeah, I noticed it after I manually updated the `def` file! It would be a little tricky though, because some names are formatted differently. For example, I use `Back Slash` instead of `Solidus` for expected character name for clarity. Would you prefer I change them in the `txt` file or perform ad-hoc substations in the script file? Or would you prefer I keep the mappings as-is?

**Comment by rintaro** [comment]
I don't think it's a good idea to update `confusables.txt` because it's from [unicode.org](https://www.unicode.org/Public/security/9.0.0/confusables.txt). 
Ad-hoc substitutions sounds good.

Lastest `confusables.txt` is probably https://www.unicode.org/Public/security/13.0.0/confusables.txt

**Comment by rintaro** [comment]
Could you add some warning text in the generated file like:
```
////////////////////////////////////////////////////////////////////////////////
// WARNING: This file is manually generated from
// utils/UnicodeData/confusables.txt and should not be directly modified.
// Run utils/generate_harness/generate_confusables.py to regenerate this file.
////////////////////////////////////////////////////////////////////////////////
```

**Comment by theblixguy** [comment]
Yeah, I will add that!

**Comment by theblixguy** [comment]
@rintaro I have updated the script and regenerated the `def` file.

File: lib/Parse/Confusables.cpp [hunk 1]
```diff
@@ -14,9 +14,22 @@
...
+swift::confusable::getConfusableAndBaseCodepointNames(uint32_t codepoint) {
+  switch (codepoint) {
+#define CONFUSABLE(CONFUSABLE_POINT, CONFUSABLE_NAME, BASE_POINT, BASE_NAME)   \
+  case CONFUSABLE_POINT:                                                       \
+    return std::make_pair(CONFUSABLE_NAME, BASE_NAME);
+#include "swift/Parse/Confusables.def"
+  default:
+    return std::make_pair("", "");
+  }
+}
```
Comment (rintaro) [inline lib/Parse/Confusables.cpp:35]: Nit pick: newline

File: include/swift/Parse/Confusables.def [hunk 2]
```diff
@@ -2,127 +2,134 @@
...
 //
 // See https://swift.org/LICENSE.txt for license information
 // See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
 //
+//
+////////////////////////////////////////////////////////////////////////////////
+// WARNING: This file is manually generated from
+// utils/UnicodeData/confusables.txt and should not be directly modified.
+// Run utils/generate_confusables.py to regenerate this file.
+////////////////////////////////////////////////////////////////////////////////
```
Comment (rintaro) [inline include/swift/Parse/Confusables.def:16]: I prefer this under the `//===--- ... ---===//` line.

File: utils/generate_confusables.py [hunk 3]
```diff
@@ -79,19 +82,30 @@ def main(args=sys.argv):
...
+//
+////////////////////////////////////////////////////////////////////////////////
+// WARNING: This file is manually generated from
+// utils/UnicodeData/confusables.txt and should not be directly modified.
+// Run utils/generate_confusables.py to regenerate this file.
+////////////////////////////////////////////////////////////////////////////////
 '''
         f.write(header)
         f.write("//===----------------------------------------------------")
         f.write("------------------===//\n\n")
```
Comment (rintaro) [inline utils/generate_confusables.py:100]: I wonder why this line (and the above) is not included in `header` 🤔

File: utils/generate_confusables.py [hunk 4]
```diff
@@ -79,19 +82,30 @@ def main(args=sys.argv):
...
-        for (confused, expected) in pairs:
-            f.write("CONFUSABLE(" + confused + ", " + expected + ")\n")
+        f.write("// CONFUSABLE(CONFUSABLE_POINT, CONFUSABLE_NAME, BASE_POINT, BASE_NAME)\n\n")
+        for (confused, confusedName, expected, expectedName) in pairs:
+            # Ad-hoc substitutions for clarity
+            mappings = { "Solidus": "Back Slash", "Reverse Solidus": "Forward Slash" }
+            newExpectedName = expectedName
+            if expectedName in mappings:
+                newExpectedName = mappings[expectedName]
+            f.write("CONFUSABLE(" + confused + ", " + '"' + confusedName + '"' + ", " + expected + ", " + '"' + newExpectedName + '"' + ")\n")
```
Comment (rintaro) [inline utils/generate_confusables.py:108]: This probably fails `validation-test/Python/python_lint.swift`. Try `@swift-ci Python lint`

File: utils/UnicodeData/confusables.txt [hunk 5]
```diff
@@ -1,11 +1,11 @@
...
-# © 2016 Unicode®, Inc.
+# Date: 2020-02-13, 01:38:49 GMT
+# © 2020 Unicode®, Inc.
 # Unicode and the Unicode Logo are registered trademarks of Unicode, Inc. in the U.S. and other countries.
 # For terms of use, see http://www.unicode.org/terms_of_use.html
 #
-# Unicode Security Mechanisms for UTR #39
-# Version: 9.0.0
+# Unicode Security Mechanisms for UTS #39
+# Version: 13.0.0
```
Comment (rintaro) [inline utils/UnicodeData/confusables.txt:8]: I will confirm if we can simply update this file in our repo.

File: utils/generate_confusables.py (hunk 3 above)
Comment (theblixguy) [reply utils/generate_confusables.py:100]: I have moved it inside `header` now.

File: utils/generate_confusables.py (hunk 4 above)
Comment (theblixguy) [reply utils/generate_confusables.py:108]: I have fixed it locally, but will run CI to double-check!

File: utils/UnicodeData/confusables.txt (hunk 5 above)
Comment (theblixguy) [reply utils/UnicodeData/confusables.txt:8]: Okay!

**Comment by theblixguy** [comment]
@swift-ci please smoke test

**Comment by theblixguy** [comment]
@swift-ci python lint

**Comment by rintaro** [comment]
@theblixguy We can update `confusables.txt`. Could you fix the failing test case so we can merge?

**Comment by theblixguy** [comment]
@rintaro Great! I have updated the test.

**Comment by theblixguy** [comment]
@swift-ci please smoke test

**Comment by rintaro** [comment]
Windows: [Build Status](https://ci-external.swift.org/view/Pull%20Request/job/swift-PR-windows/5581/)

**Review by rintaro — APPROVED** [review]

**Comment by rintaro** [comment]
@swift-ci Please smoke test Windows

**Comment by theblixguy** [comment]
@swift-ci please Python lint

**Pull request merged**